base_layer/core/src/proof_of_work/monero_rx.rs

```diff
@@ -22,97 +22,323 @@
+        println!("Creating input blob");
         let input = create_input_blob(&monero)?;
-        let cache = RandomXCache::new(flags, &key)?;
+        println!("Creating cache");
+        let cache = RandomXCache::new(flags, &monero.seed_key.as_bytes())?;
+        println!("Creating dataset");
         let dataset = RandomXDataset::new(flags, &cache, 0)?;
+        println!("Creating VM");
+        // Todo revisit check achieved difficulty. I believe this is an expensive call. Thats the reason monero does not
+        // create a new seed hash every block.
         let vm = RandomXVM::new(flags, Some(&cache), Some(&dataset))?;
-        let hash = vm.calculate_hash(&input)?;
+        println!("Calculating hash");
+        let hash = vm.calculate_hash(&input.as_bytes())?;
         let scalar = U256::from_big_endian(&hash); // Big endian so the hash has leading zeroes
         let result = MAX_TARGET / scalar;
         let difficulty = result.low_u64().into();
-        Ok(difficulty)
+        println!(
+            "Checking difficulty {:?} against target {:?}",
+            difficulty, header.pow.target_difficulty
+        );
+        if difficulty >= header.pow.target_difficulty {
+            return Err(MergeMineError::ValidationError);
+        }
+        Ok((difficulty, hash))
     }
     #[cfg(not(feature = "monero_merge_mining"))]
     {
         Err(MergeMineError::HashingError)
     }
 }
 
-fn create_input_blob(_data: &MoneroData) -> Result<String, MergeMineError> {
-    // Todo deserialize monero data to create string for  randomX vm
-    // Returning an error here so that difficulty can return 0 as this is not yet implemented.
-    Err(MergeMineError::HashingError)
+fn create_input_blob(data: &MoneroData) -> Result<Vec<u8>, MergeMineError> {
+    let serialized_header = bincode::serialize(&data.header);
+    if serialized_header.is_err() {
+        return Err(MergeMineError::SerializeError);
+    }
+
+    let serialized_root_hash = bincode::serialize(&data.transaction_root);
+    if serialized_root_hash.is_err() {
+        return Err(MergeMineError::SerializeError);
+    }
+
+    let serialized_transaction_count = bincode::serialize(&data.count);
+    if serialized_transaction_count.is_err() {
+        return Err(MergeMineError::SerializeError);
+    }
+
+    let mut pre_hash_blob = serialized_header.unwrap();
+    pre_hash_blob.append(&mut serialized_root_hash.unwrap());
+    pre_hash_blob.append(&mut serialized_transaction_count.unwrap());
+    let hash_blob = Hash::hash(pre_hash_blob.as_slice());
+    Ok(hash_blob.0.to_vec())
 }
 
-fn verify_header(_header: &BlockHeader, _monero_data: &MoneroData) -> Result<(), MergeMineError> {
-    // todo
-    // verify that our header is in coinbase
-    // todo
-    // verify that coinbase is in root.
+// todo this should move towards the validators. We need db access.
+fn verify_header(header: &BlockHeader, monero_data: &MoneroData) -> Result<(), MergeMineError> {
```

Comment:
We need to do this in the validators as we need to verify the seed key,  that will need db access

Reply:
Think this has been addressed in #2061 as part of the block validation code?